File: internal/application/commit_service_test.go
package application

import (
	"context"
	"errors"
	"strings"
	"testing"

	"github.com/hector/easy-commit/internal/config"
	"github.com/hector/easy-commit/internal/domain"
	"github.com/hector/easy-commit/internal/shared"
)

// MockGitRepository is a mock implementation of domain.GitRepository for testing
type MockGitRepository struct {
	isGitRepo         bool
	hasStagedChanges  bool
	commitError       error
	lastCommitMessage string
	commitCallCount   int
	lastCommittedMsg  string
}

func NewMockGitRepository() *MockGitRepository {
	return &MockGitRepository{
		isGitRepo:        true,
		hasStagedChanges: true,
	}
}

func (m *MockGitRepository) IsGitRepository() bool {
	return m.isGitRepo
}

func (m *MockGitRepository) HasStagedChanges() bool {
	return m.hasStagedChanges
}

func (m *MockGitRepository) Commit(ctx context.Context, message string) error {
	m.commitCallCount++
	m.lastCommittedMsg = message
	return m.commitError
}

func (m *MockGitRepository) GetLastCommitMessage() (string, error) {
	if m.lastCommitMessage == "" {
		return "", errors.New("no commits")
	}
	return m.lastCommitMessage, nil
}

// MockValidator is a mock implementation of domain.CommitValidator for testing
type MockValidator struct {
	validateError error
	validateCalls int
}

func NewMockValidator() *MockValidator {
	return &MockValidator{}
}

func (m *MockValidator) Validate(ctx context.Context, commit domain.Commit) error {
	m.validateCalls++
	return m.validateError
}

func TestNewCommitService(t *testing.T) {
	cfg := config.DefaultConfig()
	logger := shared.NewLogger(shared.ERROR)
	mockRepo := NewMockGitRepository()
	mockValidator := NewMockValidator()

	service := NewCommitService(mockRepo, mockValidator, logger, cfg)

	if service == nil {
		t.Fatal("NewCommitService() returned nil")
	}

	if service.gitRepo == nil {
		t.Error("gitRepo not set")
	}

	if service.validator == nil {
		t.Error("validator not set")
	}

	if service.logger == nil {
		t.Error("logger not set")
	}

	if service.config == nil {
		t.Error("config not set")
	}
}

func TestCommitService_CreateCommit_Success(t *testing.T) {
	cfg := config.DefaultConfig()
	logger := shared.NewLogger(shared.ERROR)
	mockRepo := NewMockGitRepository()
	mockValidator := NewMockValidator()

	service := NewCommitService(mockRepo, mockValidator, logger, cfg)
	ctx := context.Background()

	commit := domain.NewCommit(&cfg.Commit)
	commit.Type = domain.TypeFeat
	commit.Description = "add new feature"

	err := service.CreateCommit(ctx, commit)

	if err != nil {
		t.Errorf("Unexpected error: %v", err)
	}

	if mockValidator.validateCalls != 1 {
		t.Errorf("Expected validator to be called once, got %d", mockValidator.validateCalls)
	}

	if mockRepo.commitCallCount != 1 {
		t.Errorf("Expected git commit to be called once, got %d", mockRepo.commitCallCount)
	}

	expectedMsg := "feat: add new feature"
	if mockRepo.lastCommittedMsg != expectedMsg {
		t.Errorf("Expected commit message %q, got %q", expectedMsg, mockRepo.lastCommittedMsg)
	}
}

func TestCommitService_CreateCommit_NotGitRepository(t *testing.T) {
	cfg := config.DefaultConfig()
	logger := shared.NewLogger(shared.ERROR)
	mockRepo := NewMockGitRepository()
	mockRepo.isGitRepo = false
	mockValidator := NewMockValidator()

	service := NewCommitService(mockRepo, mockValidator, logger, cfg)
	ctx := context.Background()

	commit := domain.NewCommit(&cfg.Commit)
	commit.Type = domain.TypeFeat
	commit.Description = "test"

	err := service.CreateCommit(ctx, commit)

	if err == nil {
		t.Error("Expected error when not in git repository")
	}

	if !strings.Contains(err.Error(), "git repository") {
		t.Errorf("Expected error about git repository, got: %v", err)
	}

	if mockRepo.commitCallCount != 0 {
		t.Error("Git commit should not have been called")
	}
}

func TestCommitService_CreateCommit_NoStagedChanges(t *testing.T) {
	cfg := config.DefaultConfig()
	logger := shared.NewLogger(shared.ERROR)
	mockRepo := NewMockGitRepository()
	mockRepo.hasStagedChanges = false
	mockValidator := NewMockValidator()

	service := NewCommitService(mockRepo, mockValidator, logger, cfg)
	ctx := context.Background()

	commit := domain.NewCommit(&cfg.Commit)
	commit.Type = domain.TypeFeat
	commit.Description = "test"

	err := service.CreateCommit(ctx, commit)

	if err == nil {
		t.Error("Expected error when no staged changes")
	}

	if !strings.Contains(err.Error(), "staged changes") {
		t.Errorf("Expected error about staged changes, got: %v", err)
	}

	if mockRepo.commitCallCount != 0 {
		t.Error("Git commit should not have been called")
	}
}

func TestCommitService_CreateCommit_ValidationFails(t *testing.T) {
	cfg := config.DefaultConfig()
	logger := shared.NewLogger(shared.ERROR)
	mockRepo := NewMockGitRepository()
	mockValidator := NewMockValidator()
	mockValidator.validateError = shared.ErrEmptyDescription

	service := NewCommitService(mockRepo, mockValidator, logger, cfg)
	ctx := context.Background()

	commit := domain.NewCommit(&cfg.Commit)
	commit.Type = domain.TypeFeat
	commit.Description = ""

	err := service.CreateCommit(ctx, commit)

	if err == nil {
		t.Error("Expected validation error")
	}

	if mockValidator.validateCalls != 1 {
		t.Errorf("Expected validator to be called once, got %d", mockValidator.validateCalls)
	}

	if mockRepo.commitCallCount != 0 {
		t.Error("Git commit should not have been called after validation failure")
	}
}

func TestCommitService_CreateCommit_GitCommitFails(t *testing.T) {
	cfg := config.DefaultConfig()
	logger := shared.NewLogger(shared.ERROR)
	mockRepo := NewMockGitRepository()
	mockRepo.commitError = errors.New("git commit failed")
	mockValidator := NewMockValidator()

	service := NewCommitService(mockRepo, mockValidator, logger, cfg)
	ctx := context.Background()

	commit := domain.NewCommit(&cfg.Commit)
	commit.Type = domain.TypeFeat
	commit.Description = "test"

	err := service.CreateCommit(ctx, commit)

	if err == nil {
		t.Error("Expected git commit error")
	}

	if !strings.Contains(err.Error(), "git commit failed") {
		t.Errorf("Expected git error, got: %v", err)
	}

	if mockRepo.commitCallCount != 1 {
		t.Errorf("Expected git commit to be attempted, got %d calls", mockRepo.commitCallCount)
	}
}

func TestCommitService_CreateCommit_WithScope(t *testing.T) {
	cfg := config.DefaultConfig()
	logger := shared.NewLogger(shared.ERROR)
	mockRepo := NewMockGitRepository()
	mockValidator := NewMockValidator()

	service := NewCommitService(mockRepo, mockValidator, logger, cfg)
	ctx := context.Background()

	commit := domain.NewCommit(&cfg.Commit)
	commit.Type = domain.TypeFix
	commit.Description = "resolve authentication bug"
	commit.Scope = "auth"

	err := service.CreateCommit(ctx, commit)

	if err != nil {
		t.Errorf("Unexpected error: %v", err)
	}

	expectedMsg := "fix(auth): resolve authentication bug"
	if mockRepo.lastCommittedMsg != expectedMsg {
		t.Errorf("Expected commit message %q, got %q", expectedMsg, mockRepo.lastCommittedMsg)
	}
}

func TestCommitService_CreateCommit_WithBody(t *testing.T) {
	cfg := config.DefaultConfig()
	logger := shared.NewLogger(shared.ERROR)
	mockRepo := NewMockGitRepository()
	mockValidator := NewMockValidator()

	service := NewCommitService(mockRepo, mockValidator, logger, cfg)
	ctx := context.Background()

	commit := domain.NewCommit(&cfg.Commit)
	commit.Type = domain.TypeFeat
	commit.Description = "add new feature"
	commit.Body = "This is a detailed explanation of the feature"

	err := service.CreateCommit(ctx, commit)

	if err != nil {
		t.Errorf("Unexpected error: %v", err)
	}

	if !strings.Contains(mockRepo.lastCommittedMsg, "feat: add new feature") {
		t.Error("Commit message missing header")
	}

	if !strings.Contains(mockRepo.lastCommittedMsg, "This is a detailed explanation") {
		t.Error("Commit message missing body")
	}
}

func TestCommitService_CreateCommit_BreakingChange(t *testing.T) {
	cfg := config.DefaultConfig()
	logger := shared.NewLogger(shared.ERROR)
	mockRepo := NewMockGitRepository()
	mockValidator := NewMockValidator()

	service := NewCommitService(mockRepo, mockValidator, logger, cfg)
	ctx := context.Background()

	commit := domain.NewCommit(&cfg.Commit)
	commit.Type = domain.TypeFeat
	commit.Description = "change API structure"
	commit.Breaking = true
	commit.Body = "Restructured the API"

	err := service.CreateCommit(ctx, commit)

	if err != nil {
		t.Errorf("Unexpected error: %v", err)
	}

	if !strings.Contains(mockRepo.lastCommittedMsg, "feat!:") {
		t.Error("Breaking change marker '!' missing")
	}

	if !strings.Contains(mockRepo.lastCommittedMsg, "BREAKING CHANGE:") {
		t.Error("BREAKING CHANGE footer missing")
	}
}

func TestCommitService_PreviewCommit(t *testing.T) {
	cfg := config.DefaultConfig()
	logger := shared.NewLogger(shared.ERROR)
	mockRepo := NewMockGitRepository()
	mockValidator := NewMockValidator()

	service := NewCommitService(mockRepo, mockValidator, logger, cfg)

	commit := domain.NewCommit(&cfg.Commit)
	commit.Type = domain.TypeFeat
	commit.Description = "add preview feature"
	commit.Scope = "ui"

	preview := service.PreviewCommit(commit)

	expectedMsg := "feat(ui): add preview feature"
	if preview != expectedMsg {
		t.Errorf("Expected preview %q, got %q", expectedMsg, preview)
	}
}

func TestCommitService_GetConfig(t *testing.T) {
	cfg := config.DefaultConfig()
	logger := shared.NewLogger(shared.ERROR)
	mockRepo := NewMockGitRepository()
	mockValidator := NewMockValidator()

	service := NewCommitService(mockRepo, mockValidator, logger, cfg)

	returnedCfg := service.GetConfig()

	if returnedCfg == nil {
		t.Fatal("GetConfig() returned nil")
	}

	if returnedCfg.Commit.MaxDescriptionLength != cfg.Commit.MaxDescriptionLength {
		t.Error("GetConfig() returned different config")
	}
}

func TestCommitService_CreateCommit_CompleteFlow(t *testing.T) {
	cfg := config.DefaultConfig()
	logger := shared.NewLogger(shared.ERROR)
	mockRepo := NewMockGitRepository()
	mockValidator := NewMockValidator()

	service := NewCommitService(mockRepo, mockValidator, logger, cfg)
	ctx := context.Background()

	commit := domain.NewCommit(&cfg.Commit)
	commit.Type = domain.TypeRefactor
	commit.Description = "improve performance"
	commit.Scope = "core"
	commit.Body = "Optimized algorithms for better speed"
	commit.Breaking = false

	err := service.CreateCommit(ctx, commit)

	if err != nil {
		t.Errorf("Unexpected error: %v", err)
	}

	// Verify the complete flow
	if mockValidator.validateCalls != 1 {
		t.Error("Validation should have been called")
	}

	if mockRepo.commitCallCount != 1 {
		t.Error("Git commit should have been called")
	}

	// Verify complete message format
	msg := mockRepo.lastCommittedMsg
	if !strings.Contains(msg, "refactor(core): improve performance") {
		t.Error("Message missing proper header")
	}
	if !strings.Contains(msg, "Optimized algorithms") {
		t.Error("Message missing body")
	}
}

func TestCommitService_CreateCommit_AllCommitTypes(t *testing.T) {
	cfg := config.DefaultConfig()
	logger := shared.NewLogger(shared.ERROR)
	commitTypes := domain.CommitTypes{}.GetDefault()

	for _, commitType := range commitTypes {
		t.Run(commitType.Name, func(t *testing.T) {
			mockRepo := NewMockGitRepository()
			mockValidator := NewMockValidator()
			service := NewCommitService(mockRepo, mockValidator, logger, cfg)
			ctx := context.Background()

			commit := domain.NewCommit(&cfg.Commit)
			commit.Type = commitType
			commit.Description = "test " + commitType.Name

			err := service.CreateCommit(ctx, commit)

			if err != nil {
				t.Errorf("Unexpected error for %s: %v", commitType.Name, err)
			}

			if !strings.HasPrefix(mockRepo.lastCommittedMsg, commitType.Name+":") {
				t.Errorf("Message should start with %s:, got: %s", commitType.Name, mockRepo.lastCommittedMsg)
			}
		})
	}
}

File: tests/unit/application/commit-service.test.ts
import { beforeEach, describe, expect, mock, test } from 'bun:test';
import { CommitService } from '../../../src/application/services/commit-service';
import { Commit } from '../../../src/domain/entities/commit';
import type { CommitValidator } from '../../../src/domain/repositories/commit-validator';
import type { GitRepository } from '../../../src/domain/repositories/git-repository';
import type { ValidationResult } from '../../../src/domain/types';
import type { Config } from '../../../src/infrastructure/config/config-loader';
import type { Logger } from '../../../src/infrastructure/logger/logger';

describe('CommitService', () => {
  let service: CommitService;
  let mockGitRepo: GitRepository;
  let mockValidator: CommitValidator;
  let mockLogger: Logger;
  let mockConfig: Config;

  beforeEach(() => {
    // Create mock implementations
    mockGitRepo = {
      isGitRepository: mock(async () => true),
      hasStagedChanges: mock(async () => true),
      commit: mock(async () => {}),
      getLastCommitMessage: mock(async () => 'Previous commit message'),
    };

    mockValidator = {
      validate: mock(async () => ({ valid: true, errors: [] })),
    };

    mockLogger = {
      debug: mock(() => {}),
      info: mock(() => {}),
      warn: mock(() => {}),
      error: mock(() => {}),
      setLevel: mock(() => {}),
    };

    mockConfig = {
      commit: {
        maxDescriptionLength: 72,
        maxBodyLength: 500,
        invalidChars: [],
      },
      timeouts: {
        validation: 5000,
        gitCommand: 5000,
        userInput: 300000,
        context: 60000,
      },
      validator: {
        workers: 4,
      },
      logger: {
        level: 'INFO',
      },
    };

    service = new CommitService(mockGitRepo, mockValidator, mockLogger, mockConfig);
  });

  describe('createCommit', () => {
    test('should successfully create a valid commit', async () => {
      const commit = new Commit('feat', 'add feature');

      await service.createCommit(commit);

      expect(mockGitRepo.isGitRepository).toHaveBeenCalled();
      expect(mockGitRepo.hasStagedChanges).toHaveBeenCalled();
      expect(mockValidator.validate).toHaveBeenCalledWith(commit);
      expect(mockGitRepo.commit).toHaveBeenCalledWith('feat: add feature');
    });

    test('should throw error if not a git repository', async () => {
      mockGitRepo.isGitRepository = mock(async () => false);
      const commit = new Commit('feat', 'add feature');

      await expect(service.createCommit(commit)).rejects.toThrow('Not a git repository');

      expect(mockGitRepo.commit).not.toHaveBeenCalled();
    });

    test('should throw error if no staged changes', async () => {
      mockGitRepo.hasStagedChanges = mock(async () => false);
      const commit = new Commit('feat', 'add feature');

      await expect(service.createCommit(commit)).rejects.toThrow('No staged changes to commit');

      expect(mockGitRepo.commit).not.toHaveBeenCalled();
    });

    test('should throw error if validation fails', async () => {
      const validationResult: ValidationResult = {
        valid: false,
        errors: ['Error 1', 'Error 2'],
      };
      mockValidator.validate = mock(async () => validationResult);
      const commit = new Commit('feat', 'add feature');

      await expect(service.createCommit(commit)).rejects.toThrow('Validation failed');

      expect(mockGitRepo.commit).not.toHaveBeenCalled();
    });

    test('should format commit message correctly', async () => {
      const commit = new Commit('feat', 'add login', 'auth', 'Add JWT authentication', true);

      await service.createCommit(commit);

      const expectedMessage = commit.format();
      expect(mockGitRepo.commit).toHaveBeenCalledWith(expectedMessage);
    });

    test('should call git commit with formatted message', async () => {
      const commit = new Commit('fix', 'resolve bug', 'core');

      await service.createCommit(commit);

      expect(mockGitRepo.commit).toHaveBeenCalledWith('fix(core): resolve bug');
    });

    test('should log workflow steps', async () => {
      const commit = new Commit('feat', 'add feature');

      await service.createCommit(commit);

      expect(mockLogger.info).toHaveBeenCalledWith(expect.stringContaining('workflow'));
      expect(mockLogger.info).toHaveBeenCalledWith(expect.stringContaining('successfully'));
    });

    test('should handle git commit errors', async () => {
      mockGitRepo.commit = mock(async () => {
        throw new Error('Git commit failed');
      });
      const commit = new Commit('feat', 'add feature');

      await expect(service.createCommit(commit)).rejects.toThrow('Git commit failed');
    });
  });

  describe('previewCommit', () => {
    test('should return formatted commit message', () => {
      const commit = new Commit('feat', 'add feature');
      const preview = service.previewCommit(commit);

      expect(preview).toBe('feat: add feature');
    });

    test('should format commit with scope', () => {
      const commit = new Commit('fix', 'resolve bug', 'api');
      const preview = service.previewCommit(commit);

      expect(preview).toBe('fix(api): resolve bug');
    });

    test('should format commit with body', () => {
      const commit = new Commit('feat', 'add feature', undefined, 'Body text');
      const preview = service.previewCommit(commit);

      expect(preview).toContain('feat: add feature');
      expect(preview).toContain('Body text');
    });

    test('should format commit with breaking change', () => {
      const commit = new Commit('feat', 'change api', undefined, undefined, true);
      const preview = service.previewCommit(commit);

      expect(preview).toContain('BREAKING CHANGE:');
    });
  });

  describe('validateCommit', () => {
    test('should return validation result', async () => {
      const commit = new Commit('feat', 'add feature');
      const result = await service.validateCommit(commit);

      expect(result.valid).toBe(true);
      expect(result.errors).toHaveLength(0);
      expect(mockValidator.validate).toHaveBeenCalledWith(commit);
    });

    test('should return validation errors', async () => {
      const validationResult: ValidationResult = {
        valid: false,
        errors: ['Error 1', 'Error 2'],
      };
      mockValidator.validate = mock(async () => validationResult);
      const commit = new Commit('feat', 'add feature');

      const result = await service.validateCommit(commit);

      expect(result.valid).toBe(false);
      expect(result.errors).toEqual(['Error 1', 'Error 2']);
    });
  });

  describe('getConfig', () => {
    test('should return configuration', () => {
      const config = service.getConfig();

      expect(config).toBe(mockConfig);
      expect(config.commit.maxDescriptionLength).toBe(72);
      expect(config.validator.workers).toBe(4);
    });
  });

  describe('isGitRepository', () => {
    test('should return true for git repository', async () => {
      const result = await service.isGitRepository();

      expect(result).toBe(true);
      expect(mockGitRepo.isGitRepository).toHaveBeenCalled();
    });

    test('should return false for non-git directory', async () => {
      mockGitRepo.isGitRepository = mock(async () => false);

      const result = await service.isGitRepository();

      expect(result).toBe(false);
    });
  });

  describe('hasStagedChanges', () => {
    test('should return true when changes are staged', async () => {
      const result = await service.hasStagedChanges();

      expect(result).toBe(true);
      expect(mockGitRepo.hasStagedChanges).toHaveBeenCalled();
    });

    test('should return false when no changes are staged', async () => {
      mockGitRepo.hasStagedChanges = mock(async () => false);

      const result = await service.hasStagedChanges();

      expect(result).toBe(false);
    });
  });

  describe('getLastCommitMessage', () => {
    test('should return last commit message', async () => {
      const message = await service.getLastCommitMessage();

      expect(message).toBe('Previous commit message');
      expect(mockGitRepo.getLastCommitMessage).toHaveBeenCalled();
    });

    test('should handle empty commit history', async () => {
      mockGitRepo.getLastCommitMessage = mock(async () => '');

      const message = await service.getLastCommitMessage();

      expect(message).toBe('');
    });
  });

  describe('integration scenarios', () => {
    test('should complete full workflow successfully', async () => {
      const commit = new Commit(
        'feat',
        'add authentication',
        'auth',
        'Implement JWT with refresh tokens',
        false
      );

      await service.createCommit(commit);

      // Verify workflow steps
      expect(mockGitRepo.isGitRepository).toHaveBeenCalled();
      expect(mockGitRepo.hasStagedChanges).toHaveBeenCalled();
      expect(mockValidator.validate).toHaveBeenCalled();
      expect(mockGitRepo.commit).toHaveBeenCalled();

      // Verify final message format
      const expectedMessage = 'feat(auth): add authentication\n\nImplement JWT with refresh tokens';
      expect(mockGitRepo.commit).toHaveBeenCalledWith(expectedMessage);
    });

    test('should handle validation errors gracefully', async () => {
      mockValidator.validate = mock(async () => ({
        valid: false,
        errors: ['Description too long', 'Invalid characters'],
      }));
      const commit = new Commit('feat', 'add feature');

      await expect(service.createCommit(commit)).rejects.toThrow();

      expect(mockLogger.error).toHaveBeenCalled();
      expect(mockGitRepo.commit).not.toHaveBeenCalled();
    });
  });
});
